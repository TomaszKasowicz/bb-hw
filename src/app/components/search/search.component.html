<input class="flex-grow-1" type="text" [(ngModel)]="searchText" placeholder="Search By Typing" (keyup)="emitChange($event)">
<button [class.hidden]="!searchText" (click)="clear()">X</button>