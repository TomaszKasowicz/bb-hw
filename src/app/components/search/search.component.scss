:host {
  display: flex;
}

input {
  padding-right: 10px;
}

button {
  margin-left: -5px;
  background-color: var(--color-background-primary);
  border: none;
  border-bottom: 1px solid var(--color-line);
  color: var(--color-text-postfix);

  &.hidden {
    visibility: hidden;
  }
  &:hover {
    cursor: pointer;
  }
}
