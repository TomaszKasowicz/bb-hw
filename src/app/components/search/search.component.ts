import { Component, EventEmitter, OnInit, Output } from '@angular/core';

@Component({
  selector: 'app-search',
  templateUrl: './search.component.html',
  styleUrls: ['./search.component.scss']
})
export class SearchComponent implements OnInit {
  @Output()
  public searchTextChanged = new EventEmitter<string>();
  public searchText: string;

  constructor() {}

  ngOnInit(): void {}

  emitChange(event?: KeyboardEvent) {
    if (event?.key === 'Escape') {
      this.searchText = '';
    }
    this.searchTextChanged.emit(this.searchText);
  }

  clear() {
    this.searchText = '';
    this.emitChange();
  }
}
