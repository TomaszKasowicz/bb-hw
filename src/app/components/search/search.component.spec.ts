import { DebugElement } from '@angular/core';
import { FormsModule } from '@angular/forms';
import { By } from '@angular/platform-browser';
import { Spectator, createKeyboardEvent } from '@ngneat/spectator';
import { createComponentFactory } from '@ngneat/spectator/jest';
import { SearchComponent } from './search.component';

describe('SearchComponent', () => {
  let spectator: Spectator<SearchComponent>;

  const createComponent = createComponentFactory({
    component: SearchComponent,
    imports: [FormsModule]
  });

  beforeEach(() => {
    spectator = createComponent();
  });

  it('should create', () => {
    expect(spectator.component).toBeTruthy();
  });

  it('should apply hidden class to X button when searchText is empty', () => {
    spectator.detectChanges();
    const buttonElem = spectator.query('button');
    expect(spectator.fixture).toMatchSnapshot();
    expect(buttonElem).toHaveClass('hidden');
  });

  it('should show X button when searchText is not falsy', () => {
    spectator.component.searchText = 'abc';
    spectator.detectChanges();

    const buttonElem = spectator.query('button');
    expect(spectator.fixture).toMatchSnapshot();
    expect(buttonElem).not.toHaveClass('hidden');
  });

  it('should emit change when input is filled', done => {
    const input = spectator.fixture.debugElement.query(By.css('input'));
    input.nativeElement.value = 'test';
    input.nativeElement.dispatchEvent(new Event('input'));
    spectator.detectChanges();

    expect(spectator.fixture).toMatchSnapshot();
    spectator.component.searchTextChanged.subscribe(value => {
      expect(value).toEqual('test');
      done();
    });
    spectator.dispatchKeyboardEvent(input, 'keyup', 't');
  });

  it('should clear searchText when Escape pressed', done => {
    spectator.component.searchText = 'test';
    spectator.detectChanges();
    const input = spectator.fixture.debugElement.query(By.css('input'));

    spectator.component.searchTextChanged.subscribe(value => {
      expect(value).toEqual('');
      done();
    });

    spectator.dispatchKeyboardEvent(input, 'keyup', 'Escape');
  });

  it('should clear searchText when X is clicked', done => {
    spectator.component.searchText = 'test';
    spectator.detectChanges();

    spectator.component.searchTextChanged.subscribe(value => {
      expect(value).toEqual('');
      done();
    });
    spectator.click('button');
  });
});
