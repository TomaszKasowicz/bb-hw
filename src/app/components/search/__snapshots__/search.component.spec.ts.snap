// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchComponent should apply hidden class to X button when searchText is empty 1`] = `
<app-search
  __ngContext__={[Function LRootView]}
  searchTextChanged={[Function EventEmitter_]}
>
  <input
    class="flex-grow-1 ng-untouched ng-pristine ng-valid"
    placeholder="Search By Typing"
    type="text"
  /><button
    class="hidden"
  >
    X
  </button>
</app-search>
`;

exports[`SearchComponent should emit change when input is filled 1`] = `
<app-search
  __ngContext__={[Function LRootView]}
  searchText={[Function String]}
  searchTextChanged={[Function EventEmitter_]}
>
  <input
    class="flex-grow-1 ng-untouched ng-valid ng-dirty"
    placeholder="Search By Typing"
    type="text"
  /><button>
    X
  </button>
</app-search>
`;

exports[`SearchComponent should show X button when searchText is not falsy 1`] = `
<app-search
  __ngContext__={[Function LRootView]}
  searchText={[Function String]}
  searchTextChanged={[Function EventEmitter_]}
>
  <input
    class="flex-grow-1 ng-untouched ng-pristine ng-valid"
    placeholder="Search By Typing"
    type="text"
  /><button>
    X
  </button>
</app-search>
`;
